flake8_trio/visitors/visitor91x.py

```diff
@@ -204,123 +340,154 @@ def visit_Try(self, node: ast.Try):
 
-            self.visit_nodes(node.finalbody)
-            self.uncheckpointed_statements.difference_update(added)
+    def leave_Try(self, original_node: cst.Try, updated_node: cst.Try) -> cst.Try:
+        self.restore_state(original_node)
+        return updated_node
 
-    # valid checkpoint if both body and orelse checkpoint
-    def visit_If(self, node: ast.If | ast.IfExp):
+    def leave_If_test(self, node: cst.If | cst.IfExp) -> None:
         if not self.async_function:
             return
-        # visit condition
-        self.visit_nodes(node.test)
-        outer = self.uncheckpointed_statements.copy()
+        self.save_state(node, "uncheckpointed_statements", copy=True)
 
-        # visit body
-        self.visit_nodes(node.body)
-        body_outer = self.uncheckpointed_statements
-
-        # reset to after condition and visit orelse
-        self.uncheckpointed_statements = outer
-        self.visit_nodes(node.orelse)
+    def leave_If_body(self, node: cst.If | cst.IfExp) -> None:
+        if not self.async_function:
+            return
 
-        # union of both branches is the new set of unhandled entries
-        self.uncheckpointed_statements.update(body_outer)
+        # restore state to after test, saving current state instead
+        (
+            self.uncheckpointed_statements,
+            self.outer[node]["uncheckpointed_statements"],
+        ) = (
+            self.outer[node]["uncheckpointed_statements"],
+            self.uncheckpointed_statements,
+        )
 
-    # inline if
-    visit_IfExp = visit_If
+    def leave_If(self, original_node: cst.If, updated_node: cst.If) -> cst.If:
+        if self.async_function:
+            # merge current state with post-body state
+            self.uncheckpointed_statements.update(
+                self.outer[original_node]["uncheckpointed_statements"]
+            )
+        return updated_node
+
+    # libcst calls attributes in the order they appear in the code, so we manually
+    # rejig the order here
+    def visit_IfExp(self, node: cst.IfExp) -> bool:
+        _ = node.test.visit(self)
+        self.leave_If_test(node)
+        _ = node.body.visit(self)
+        self.leave_If_body(node)
+        _ = node.orelse.visit(self)
+        self.leave_If(node, node)  # type: ignore
+        return False
+
+    def visit_While(self, node: cst.While | cst.For):
+        self.save_state(
+            node,
+            "loop_state",
+            copy=True,
+        )
+        self.loop_state = LoopState()
+        self.infinite_loop = self.body_guaranteed_once = False
 
     # Check for yields w/o checkpoint in between due to entering loop body the first time,
     # after completing all of loop body, and after any continues.
     # yield in else have same requirement
     # state after the loop same as above, and in addition the state at any break
-    def visit_loop(self, node: ast.While | ast.For | ast.AsyncFor):
```

Comment:
visit_loop now also split into a lot of different visit/leavers, it retains a lot of logic but changes the logic as previously mentioned from doing two passes of the loop body (first one to calculate what state the loop can be in when it starts, the second one to actually raise errors depending on that state), to injecting a special statement into `uncheckpointed_statements` that makes errors raised become "artificial errors", and depending on what the potential end-state of the loop is we either toss those out or turn them into actual errors.